text/006-local-js-dependencies.md

```diff
@@ -0,0 +1,347 @@
+  loadable inside the web browser, but otherwise having a similar interface to
+  `--no-modules` today, detailed below.
+
+This RFC proposes rethinking these output modes as follows:
+
+| Target Environment      | CLI Flag    | Module Format | User Experience                          | How are Local JS Snippets Loaded?                                                            |
+|-------------------------|-------------|---------------|------------------------------------------|----------------------------------------------------------------------------------------------|
+| Node.js without bundler | `--nodejs`  | Common.js     | `require()` the main JS glue file        | Main JS glue file `require()`s crates' local JS snippets.                                    |
+| Web without bundler     | `--browser` | ES Modules    | `<script>` pointing to main JS glue file, using `type=module` | `import` statements cause additional network requests for crates' local snippets.            |
+| Web with bundler        | none        | ES Modules    | `<script>` pointing to main JS glue file | Bundler links crates' local snippets into main JS glue file. No additional network requests except for the `wasm` module itself. |
+
+It is notable that browser with and without bundler is almost the same as far
+as `wasm-bindgen` is concerned: the only difference is that if we assume a
+bundler, we can rely on the bundler polyfilling wasm-as-ES-module for us.
+Note the `--browser` here is relatively radically different today and as such
+would be a breaking change. It's thought that the usage of `--browser` is small
+enough that we can get away with this, but feedback is always welcome on this
+point!
+
+The `--no-modules` flag doesn't really fit any more as the `--browser` use case
+is intended to subsume that. Note that the this RFC proposes only having the
+bundler-oriented and browser-oriented modes supporting local JS snippets for
+now, while paving a way forward to eventually support local JS snippets in
+Node.js. The `--no-modules` could eventually also be supported in the same
+manner as Node.js is (once we're parsing the JS file and rewriting the exports),
+but it's proposed here to generally move away from `--no-modules` towards
+`--browser`.
+
+
+The `--browser` output is currently considered to export an initialization
+function which, after called and the returned promise is resolved (like
+`--no-modules` today) will cause all exports to work when called. Before the
+promise resolves all exports will throw an error when called.
+
+### JS files depending on other JS files
+
+One tricky point about this RFC is when a local JS snippet depends on other JS
+files. For example your JS might look like:
+
+```js
+// js/foo.js
+
+import { foo } from '@some/npm-package';
+import { bar } from './bar.js'
+
+// ...
+```
+
+As designed above, these imports would not work. It's intended that we
```

Comment:
I had covered this extensively several months ago:

https://github.com/rustwasm/team/issues/92#issuecomment-374898503
https://github.com/rustwasm/team/issues/36#issuecomment-362454792
https://github.com/rustwasm/team/issues/36#issuecomment-364712488

You don't need to read them, I'll summarize it here.

Basically, wasm-bindgen would look through the crate and find all files that end in `.js`. Then it would put those files into the custom section (with the file path relative to the crate).

When generating the output, it would then recreate the same folder structure and place the `.js` files into the output folder.

This has a few major advantages:

* wasm-bindgen doesn't need to parse `.js` files at all. It only needs to read the file and keep track of the relative file path.

* It fully supports JS files importing other JS files, without any effort from wasm-bindgen (we get it "for free").

* It fully supports dynamic `import()`.

   The only other way to support dynamic `import()` is to explicitly list the `.js` files in metadata (such as in `package.json` or `Cargo.toml`).

   But with this approach, everything "Just Works", without needing to specify any metadata.

Reply:
> I think we'd still need a parser to support --no-modules and --nodejs, so while it may mean that some use cases don't need a parser in general I think we'd still need it.

That is true, that negates one of the advantages. Though I think the other advantages are still really good!

> Incremental builds in Rust I don't think would be supported well.

Oh, I hadn't considered that. That's a good point. If users manually specified the `.js` files in `package.json`, would those changes be picked up? Or would it only be changes in `Cargo.toml` that are picked up?

> I'd be worried about namespacing issues here as well.

I had put a lot of thought into that (several months ago), and I had a few solutions for that. They basically all involved putting each crate into a different subfolder, thus avoiding the namespace issue entirely. So, similar to your `/$hash/js/foo.js` idea.

As for absolute imports, that's not a problem: ES6 modules and bundlers already don't support absolute imports (or at least, support them very poorly). And in order to publish your packages, you can't use absolute imports anyways. So there's no issue with mandating file-relative or crate-relative imports.

> To actually do this in the macro I think we'd need to do something like [...]

I hadn't thought too much about the actual implementation in wasm-bindgen. It does seem rather tricky, since you can't have top-level function calls.

And given that it has issues with incremental compilation, maybe we should consider other alternatives instead.

The only alternative I know of is to manually list `.js` files in `package.json` or `Cargo.toml`.

Reply:
I think that we'll likely end up in a position where you have to list the JS files to be included, but I like the idea of automatically namespacing everything and preserving file structure to support things like dynamic import. I don't think we'd want to disallow static absolute paths if we force Rust to use absolute paths for now, but with a JS parser we could rewrite to absolute paths (or at least with our hash prefix) and force dynamic imports to never use absolute paths.

I do really like the UI though of "just pull in all JS", and I'm wondering if we should add a feature to Cargo to support this. It wouldn't be too too hard to say "Cargo I depend on this entire directory, anything added or changed here please rerun me" and that could be used to easily include a `js` folder in a wasm-bindgen application

Reply:
Perhaps we're using a different meaning for "absolute path"?

To me, "absolute path" means an absolute path on the user's harddrive, which isn't acceptable since the filepath is outside of the crate (and thus it will break for anybody other than that specific user). Perhaps you mean something different?

Absolute paths are basically never used (for the above reason), instead relative paths are the norm, so what do you mean by "we could rewrite to absolute paths"?

> force dynamic imports to never use absolute paths

How can that be enforced? Dynamic imports are resolved at runtime, so it's impossible (in the general case) to figure out what file they are importing.

That's why the "slurp up all the `.js` files" (or "list all the `.js` files in `package.json`") strategy is necessary.

> It wouldn't be too too hard to say "Cargo I depend on this entire directory, anything added or changed here please rerun me" and that could be used to easily include a js folder in a wasm-bindgen application

That sounds like it could be a useful feature even outside of wasm-bindgen!

Reply:
Oh sorry, yes we are using the same term for different things! This is about local files importing other local files, so I'd imagine that we do something like this:

```js
import { foo } from './other-local.js'; // allowed, it's relative
import { bar } from '../other-local.js'; // allowed, it's relative
import { baz } from '/other-local.js'; // allowed, but we have to rewrite this import


import('./x.js'); // we'll make sure this works
import('/x.js'); // this will always fail at runtime, no way we can make this work
```

In Rust code we'd have:

```rust
#[wasm_bindgen(module = "./foo.js")] // disallowed, don't know how to implement today
extern {}

#[wasm_bindgen(module = "../foo.js")] // disallowed, don't know how to implement today
extern {}

#[wasm_bindgen(module = "/foo.js")] // allowed, rooted-at-the-crate-root absolute path
extern {}
```

By "absolute" I mean "absolute but rooted at the crate root". 

I think how this all works though is somewhat orthogonal to whether we slurp up JS files or not. On one axis we have what all the paths are, and on another axis we have "do you list the files to include or do we automatically include?". I think on "how do handle paths" it's relatively unambiguous about what to do perhaps modulo minor tweaks. For "do we slurp up JS or not?" I'm proposing that we don't do it yet, but we leave ourselves an option in the future for doing so. (but also not doing a list-some-files-to-include for now)

Reply:
I had assumed that `/` paths simply wouldn't be allowed in `.js` files (only `./` and `../` would be allowed).

I think that it would be confusing and inconsistent to have different behavior with `import` statements and `import()` expression.

Also, using `/` in ES6 import paths is basically never done (and with browsers it means "relative to the web server root", not "relative to the crate root"), another reason to not use `/` in `.js` files.

> By "absolute" I mean "absolute but rooted at the crate root".

Okay, glad we're on the same page now. I would personally call those "crate paths" or something. Maybe we should bikeshed the name a little bit, since it will show up in the docs?

> For "do we slurp up JS or not?" I'm proposing that we don't do it yet, but we leave ourselves an option in the future for doing so. (but also not doing a list-some-files-to-include for now)

Sure, that's fine. It's only needed for dynamic `import()`, which we will need to support eventually, but not in the MVP.

Reply:
Hm ok sorry, I may not be being clear. I'll try to respond to those points:

* We must currently require absolute p
* Yes `import { ... }` and `import(...)` is inconsistent because in the limit `import { ... }` allows absolute paths

Reply:
Hm ok sorry, I may not be being clear. I'll try to respond to those points:

* We will eventually allow absolute and relative paths in `#[wasm_bindgen(module = "path")]`
  * Absolute paths are rooted at the crate root
  * Relative paths cannot be implemented with stable Rust today. We'll enable this in the future.
* `import { ... }` will, like `#[wasm_bindgen]`, allow all three paths
  * The RFC as proposed right now, however, says we'll be doing this at a future date and won't implement local-js-depending-on-local-js just yet. It just won't work and will fail with a weird error at bundler/load time
  * Eventually, absolute paths are intended to work and will be rooted at the crate root
  * Eventually, relative paths will also work as we'll have a way to ship multiple files preserving filesystem structure (bit it declarative as to which files are shipped or inferred)
* `import(...)` will disallow absolute paths.
  * Like `import { ... }` this won't work to start off with at all and will fail with weird errors
  * Like `import { ... }` relative paths will work by shipping multiple files
  * Absolute paths won't ever work because we'll have to rewrite the filesystem hierarchy to handle namespacing conflicts. 

Sorry this is so confusing, I'm trying to both lay out a plan for what we should do *now* in this RFC with wasm-bindgen while also handle all of what you're saying and plan for the eventual support for all these features. I think that a lot is getting lost in the middle as it's not clear what phase of development is being discussed.

For absolute paths and/or crate paths, we may want to figure out a different syntax if it's very rarely used in the rest of the JS ecosystem. We could try out like `$crate/...` perhaps? My point though about allowing it in JS files (not in `import(..)`, only in static imports) is that we're requiring it in Rust code (for now) so it seems natural to write in JS code too.

Reply:
> We will eventually allow absolute and relative paths in #[wasm_bindgen(module = "path")]

We are in agreement here.

> import { ... } will, like #[wasm_bindgen], allow all three paths 

We are in disagreement here. I don't think `.js` files should ever support absolute/crate paths, only relative paths.

I think the inconsistency between `import` statements and the `import()` expression isn't good. The spec authors put a lot of effort into making `import()` equivalent to `import`, and for good reason.

And I think that we shouldn't break the already existing conventions for `.js` files (which are exclusively relative paths).

> My point though about allowing it in JS files (not in import(..), only in static imports) is that we're requiring it in Rust code (for now) so it seems natural to write in JS code too.

I understand that viewpoint, I just don't agree with it.

I think that `.js` files are in a sense a "different world" from `.rs` files, so using different path systems (crate vs relative) is okay.

Especially because relative paths in `.rs` files won't work anytime soon (and might *never* work), so there's already an inconsistency between `.rs` and `.js`.

It certainly would be nice if we *could* make the path system consistent between Rust and JS, but given the existence of `import()` and the limitations of proc-macro, it doesn't seem possible.